### ISSUE TITLE
Time zones are fun they said.

### ISSUE THREAD

**Issue by GesJeremie** [workarea-commerce/workarea#278]
## Describe the bug
Changing the time zone of the build results in a failing build.

## To Reproduce

Tweak the time_zone configuration:

```ruby
# application.rb
module MyAwesomeStore
  class Application < Rails::Application
    # Initialize configuration defaults for originally generated Rails version.
    config.load_defaults 5.1
    config.eager_load_paths << "#{config.root}/lib/ext/freedom_patches"

    config.time_zone = 'Sydney'
    # ....
  end
end
```

Run the tests.

Below are the tests failing:

- Workarea::Admin::Dashboards::InsightsGraphsTest#test_decreasing 
- Workarea::Admin::Dashboards::InsightsGraphsTest#test_increasing
- Workarea::Admin::Dashboards::InsightsGraphsTest#test_time_frames 
- Workarea::Admin::Dashboards::InsightsGraphsTest#test_starting_with_nothing
- Workarea::Admin::Dashboards::InsightsGraphsTest#test_no_change
- Workarea::Admin::Dashboards::MarketingViewModelTest#test_email_signups
- Workarea::Admin::Dashboards::PeopleViewModelTest#test_new_vs_returning
- Workarea::Admin::Reports::AverageOrderValueViewModelTest#test_graph_data
- Workarea::Admin::Reports::SalesOverTimeViewModelTest#test_graph_data
- Workarea::Admin::ReportsSystemTest#test_first_time_vs_returning_sales_test
- Workarea::Admin::ReportsSystemTest#test_sales_over_time
- Workarea::Admin::ReportsSystemTest#test_average_order_value
- Workarea::Insights::CustomerAcquisitionTest#test_results
- Workarea::Metrics::ProductForLastWeekTest#test_aggregate_last_week!
- Workarea::Metrics::SearchForLastWeekTest#test_aggregate_last_week!
- Workarea::Reports::AverageOrderValueTest#test_date_ranges
- Workarea::Reports::AverageOrderValueTest#test_quarter
- Workarea::Reports::AverageOrderValueTest#test_month
- Workarea::Reports::AverageOrderValueTest#test_year
- Workarea::Reports::AverageOrderValueTest#test_by_day_of_week
- Workarea::Reports::AverageOrderValueTest#test_by_day
- Workarea::Reports::AverageOrderValueTest#test_week
- Workarea::Reports::FirstTimeVsReturningSalesTest#test_by_day_of_week
- Workarea::Reports::FirstTimeVsReturningSalesTest#test_year
- Workarea::Reports::FirstTimeVsReturningSalesTest#test_week
- Workarea::Reports::FirstTimeVsReturningSalesTest#test_by_day
- Workarea::Reports::FirstTimeVsReturningSalesTest#test_date_ranges
- Workarea::Reports::FirstTimeVsReturningSalesTest#test_quarter
- Workarea::Reports::FirstTimeVsReturningSalesTest#test_month
- Workarea::Reports::SalesByCategoryTest#test_date_ranges
- Workarea::Reports::SalesByCountryTest#test_date_ranges
- Workarea::Reports::SalesByDiscountTest#test_date_ranges
- Workarea::Reports::SalesByMenuTest#test_date_ranges
- Workarea::Reports::SalesByProductTest#test_date_ranges
- Workarea::Reports::SalesBySkuTest#test_date_ranges
- [and few others ...]

## Expected behavior
Changing the time zone of the build should not result in a failing build.

## Workarea Setup
I've been able to reproduce the bug on a vanilla Workarea app (v3.4.19)

## Additional context
It seems most of the failures are related to the Metrics / Reports system. After some investigation I believe one of the potential problems is due to how we query the Reports.

The metrics seem to be stored as UTC (how it should be). But it seems we are querying the Reports for the current timezone but not converted as UTC before hand.

For instance:

```ruby
# /workarea-core-3.4.19/app/queries/workarea/reports/sales_by_product.rb
module Workarea
  module Reports
    class SalesByProduct
      include Report

      self.reporting_class = Metrics::ProductByDay
      self.sort_fields = %w(units_sold orders merchandise discounts tax revenue)

      def aggregation
        [filter, project_used_fields, group_by_product]
      end

      def filter
        {
          '$match' => {
            'reporting_on' => { '$gte' => starts_at, '$lte' => ends_at },
            '$or' => [
              { 'orders' => { '$gt' => 0 } },
              { 'units_sold' => { '$gt' => 0 } }
            ]
          }
        }
      end

      def project_used_fields
        {
          '$project' => {
            'product_id' => 1,
            'orders' => 1,
            'units_sold' => 1,
            'merchandise' => 1,
            'discounts' => 1,
            'tax' => 1,
            'revenue' => 1
          }
        }
      end

      def group_by_product
        {
          '$group' => {
            '_id' => '$product_id',
            'orders' => { '$sum' => '$orders' },
            'units_sold' => { '$sum' => '$units_sold' },
            'merchandise' => { '$sum' => '$merchandise' },
            'discounts' => { '$sum' => '$discounts' },
            'tax' => { '$sum' => '$tax' },
            'revenue' => { '$sum' => '$revenue' }
          }
        }
      end
    end
  end
end
```

The filter is given a constraint **starts_at** and **ends_at** included by the following concern:

```ruby
# /workarea-core-3.4.19/app/queries/workarea/reports/report.rb
module Workarea
  module Reports
    module Report
      extend ActiveSupport::Concern

      included do
        attr_reader :params
        cattr_accessor :reporting_class, :sort_fields, :sort_directions

        self.sort_directions = %w(desc asc)
      end

      def initialize(params = {})
        @params = params.to_h.with_indifferent_access
      end

      def slug
        self.class.name.demodulize.underscore
      end

      def starts_at
        @starts_at ||= begin
          default = Workarea.config.reports_default_starts_at.call
          value = Time.zone.parse(params[:starts_at].to_s) || default rescue default
          value.beginning_of_day
        end
      end

      def ends_at
        @ends_at ||= begin
          value = Time.zone.parse(params[:ends_at].to_s) || Time.current rescue Time.current
          value.end_of_day
        end
      end

      def sort
        { '$sort' => sort_value }
      end

      def limit
        { '$limit' => Workarea.config.reports_max_results }
      end

      def sort_by
        @sort_by ||= params[:sort_by].presence_in(sort_fields) || sort_fields.first
      end

      def sort_value
        { sort_by => sort_direction == 'desc' ? -1 : 1 }
      end

      def sort_direction
        @sort_direction ||= params[:sort_direction].presence_in(sort_directions) ||
          sort_directions.first
      end

      def count
        results.length
      end

      def more_results?
        count >= Workarea.config.reports_max_results
      end

      def results
        @results ||= Rails.cache.fetch(cache_key, expires_in: Workarea.config.cache_expirations.reports) do
          reporting_class.collection.aggregate(aggregation + [sort, limit]).to_a
        end
      end

      def cache_key
        cache_params = params.reject { |k, v| k.blank? || v.blank? }.sort

        [
          :reports,
          slug,
          Time.current.strftime('%Y%m%d'),
          Digest::SHA1.hexdigest(cache_params.to_s)
        ].join('/')
      end
    end
  end
end
```

I believe **starts_at** and **ends_at** should return the  absolute value as .utc since we are querying dates added in UTC.

![image](https://user-images.githubusercontent.com/1700610/70360357-0edd3c00-1833-11ea-9fa3-6dbec78406f7.png)

**Comment by tubbo** [comment]
thank you for the potato, and the bug report! we're investigating now.

**Comment by bencrouse** [comment]
I'm closing, as this was fixed in #279
